### PR TITLE
c3 Charting Integration - Line, Bar (stacked), Pie. Can easily add more.

### PR DESCRIPTION
**Pull request by anwar6953**
Removed debug chart type. Will be replaced by Table chart type in a separate PR.

Click on the gif for a potentially larger view:

### CODE REVIEW

File: cdap-ui/app/directives/epoch/charts.js [hunk 1]
```diff
@@ -33,6 +33,7 @@ ngEpoch.controller('epochController', function ($scope, $compile, caskWindowMana
     }, options);
 
     angular.extend(options, forcedOpts || {}, {
+      // TODO: Do we need to copy $scope.history here? I think not.
```
Comment (anwar6953) [inline cdap-ui/app/directives/epoch/charts.js]: I didn't think $scope.history would have any data upon init.
Can we remove this? (Epoch chart directive).

File: cdap-ui/app/features/dashboard/templates/userdashboard.html [hunk 2]
```diff
@@ -7,7 +7,7 @@ <h3 class="text-center"> No cool widgets in this board! <a ui-sref="dashboard.us
 </div>
 
 <div class="row" ng-controller="WidgetColCtrl" ng-if="!unknownBoard">
-  <div ng-class="{ 'col-xs-12': colWidth.fullWidth, 'col-xs-12 col-md-4': colWidth.oneThird }"
+  <div ng-class="{ 'col-xs-12': colWidth.fullWidth, 'col-xs-12 col-md-6': colWidth.oneThird }"
```
Comment (anwar6953) [inline cdap-ui/app/features/dashboard/templates/userdashboard.html]: I also updated the widget widths in this PR.

**Comment by anwar6953** [comment]
I have removed the dots at each point.

File: cdap-ui/app/directives/c3/charts.js [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+            metricNames.push(thisMetric.label);
+            for (var j = 0; j < thisMetric.values.length; j++) {
+              thisData.push(thisMetric.values[j].y);
+            }
+            columns.push(thisData);
+          }
+          var xValues = ['x'];
+          for (var i = 0; i < newVal[0].values.length; i++) {
+            xValues.push(newVal[0].values[i].time);
+          }
```
Comment (elfenheart) [inline cdap-ui/app/directives/c3/charts.js]: Can you comment what each of these for-loops does?

File: cdap-ui/app/directives/c3/charts.js [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    myTooltip = {
+      format: {
+        value: function (value, ratio, id) {
+          return d3.format(',')(value);
+        }
+      }
+    }
+
+    var chartConfig = {bindto: $scope.options.el, data: data, tooltip: myTooltip};
+    chartConfig.size = { height: 200 };
```
Comment (elfenheart) [inline cdap-ui/app/directives/c3/charts.js]: Should you be getting this height from the attr.chart-height

File: cdap-ui/app/directives/c3/charts.js [hunk 5]
```diff
@@ -0,0 +1,179 @@
+var ngC3 = angular.module(PKG.name+'.commons');
+
+var baseDirective = {
+  restrict: 'E',
+  replace: true,
+  template: '<div class="c3"></div>',
+  scope: {
+    history: '=',
+    stream: '='
```
Comment (elfenheart) [inline cdap-ui/app/directives/c3/charts.js]: Do you need this stream?

File: cdap-ui/app/features/dashboard/templates/widgets/c3-bar.html [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<div class="{{wdgt.color}}" ng-controller="WidgetTimeseriesCtrl">
+
+  <c3-bar
+    data-history="chartHistory"
+    chart-height="200"
+    chart-ticks="{ bottom: 3 }"
+    chart-axes="['left', 'bottom']"
```
Comment (elfenheart) [inline cdap-ui/app/features/dashboard/templates/widgets/c3-bar.html]: these additional attributes are not actually being used in your directive.

File: cdap-ui/app/services/helpers.js [hunk 7]
```diff
@@ -96,11 +96,29 @@ angular.module(PKG.name+'.services')
...
+      minutes = minutes < 10 ? '0' + minutes : minutes;
+      seconds = seconds < 10 ? '0' + seconds : seconds;
+      // example format: 1:09:11 PM
+      return hours + ':' + minutes + ':' + seconds + ' ' + ampm;
+    }
+
+    function dateToString(e) {
+      var date = new Date(e * 1000);
+      return formatDate(date);
+    }
```
Comment (elfenheart) [inline cdap-ui/app/services/helpers.js]: Angular Moment Filter already providing us with this capability right?

Comment (anwar6953) [inline cdap-ui/app/services/helpers.js]: Can you call the Angular Moment filter like a function? and use the value in other javascript?

File: cdap-ui/app/features/dashboard/templates/widgets/c3-bar.html (hunk 6 above)
Comment (anwar6953) [inline cdap-ui/app/features/dashboard/templates/widgets/c3-bar.html]: Ah yeah, I've removed them in: https://github.com/caskdata/cdap/pull/2100

File: cdap-ui/app/directives/c3/charts.js (hunk 4 above)
Comment (anwar6953) [inline cdap-ui/app/directives/c3/charts.js]: I've done that in improvements to: https://github.com/caskdata/cdap/pull/2100

File: cdap-ui/app/directives/c3/charts.js (hunk 5 above)
Comment (anwar6953) [inline cdap-ui/app/directives/c3/charts.js]: no. removed.

File: cdap-ui/app/services/helpers.js (hunk 7 above)
Comment (elfenheart) [inline cdap-ui/app/services/helpers.js]: You should be able to. Inject $filter, then you can call $filter('amDateFormat')

http://stackoverflow.com/questions/14302267/how-to-use-a-filter-in-a-controller

File: cdap-ui/Gulpfile.js [hunk 8]
```diff
@@ -18,7 +18,9 @@ gulp.task('css:lib', ['fonts'], function() {
       './bower_components/font-awesome/css/font-awesome.min.css',
       './bower_components/epoch/epoch.min.css',
       './bower_components/ng-sortable/dist/ng-sortable.min.css',
-      './bower_components/angular-ui-select/dist/select.min.css'
+      './bower_components/angular-ui-select/dist/select.min.css',
+      './bower_components/c3/c3.min.css',
+      './bower_components/ng-sortable/dist/ng-sortable.min.css'
```
Comment (anwar6953) [inline cdap-ui/Gulpfile.js]: Why this?

File: cdap-ui/app/services/helpers.js (hunk 7 above)
Comment (anwar6953) [inline cdap-ui/app/services/helpers.js]: Ok, so this functionality was available.
Take a look how I did it:
https://github.com/caskdata/cdap/pull/2184/files#diff-397d9e155b58c0fb9056a2873515f0afR18

For some reason, $scope was available in the render call, but not $filter. So I put it on the $scope object. Any idea how to get around that?

**Pull request merged**